{"nbformat":4,"nbformat_minor":0,"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.3"},"colab":{"name":"jun_anl_exceptions.ipynb","provenance":[],"collapsed_sections":[]}},"cells":[{"cell_type":"markdown","metadata":{"id":"c2KtXmjxtgw-"},"source":["# Исключения\n","## Домашняя работа"]},{"cell_type":"markdown","metadata":{"id":"0Vhx-zKitgxA"},"source":["### Вопросы по лекциям."]},{"cell_type":"markdown","metadata":{"id":"14lleKewtgxC"},"source":["Как поймать вообще все ошибки, которые могут произойти?"]},{"cell_type":"markdown","metadata":{"id":"VQnYGkVKtgxD"},"source":["**Ответ:**\n","\n","try:\n","    \n","    ....\n","\n","except:\n","    \n","    ....\n","\n","или \n","\n","try:\n","    \n","    ....\n","\n","except Exception:\n","    \n","    ...."]},{"cell_type":"markdown","metadata":{"id":"paSc2JE6tgxD"},"source":["Сколько раз подряд можно указывать except?"]},{"cell_type":"markdown","metadata":{"id":"bU0lzJPBtgxE"},"source":["**Ответ:**\n","Столько, сколько необходимо для отлавливания всех ошибок"]},{"cell_type":"markdown","metadata":{"id":"9wQ-yXpGtgxF"},"source":["Вы хотите с помощью print вывести название ошибки в консоль, как это сделать?"]},{"cell_type":"markdown","metadata":{"id":"iE1iqRZctgxH"},"source":["**Ответ:**\n","\n","try:\n","    \n","    ....\n","\n","except Exception as exc:\n","    \n","    print(exc)\n"]},{"cell_type":"markdown","metadata":{"id":"EDAi3ZFBtgxI"},"source":["Вы хотите с помощью print вывести параметры ошибки в консоль, как это сделать?"]},{"cell_type":"markdown","metadata":{"id":"qWimgtF3tgxK"},"source":["**Ответ:**\n","\n","try:\n","    \n","    ....\n","\n","except Exception as exc:\n","    \n","    print(exc.args)"]},{"cell_type":"markdown","metadata":{"id":"leOCJIlRtgxK"},"source":["Что такое DeprecationWarning?\r\n"]},{"cell_type":"markdown","metadata":{"id":"KyaynuyvtgxL"},"source":["**Ответ:**\n","Предупреждение о том, что функционал устарел и не рекомендуется к использованию.\n"]},{"cell_type":"markdown","metadata":{"id":"oxgCUHzotgxM"},"source":["### Разминочные задания. \n","<br>\n","\n","Вам даны две функции. Поисследуйте, какие ошибки могут возникнуть при их реализации. Обработайте эти ошибки."]},{"cell_type":"code","metadata":{"id":"X4nsMUUAtgxN"},"source":["def div():\n","    for i in range(2):\n","        try:\n","            x = int(input(\"enter a number: \"))\n","            y = int(input(\"enter another number: \"))  \n","            try:\n","                print(x, '/', y, '=', x/y)\n","            except ZeroDivisionError as exc_ZDE:\n","                print(f'Деление на 0 - {exc_ZDE}')\n","        except ValueError as exc_VE:\n","            print(f'Ошибка в значении {exc_VE}')\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"hoDmt-nStgxS"},"source":["def sumOfPairs(L1, L2):\n","    sum = 0\n","    sumOfPairs = []\n","    for i in range(len(L1)):\n","        try:\n","            sumOfPairs.append(L1[i]+L2[i])\n","        except (IndexError, TypeError) as exc:\n","            print(f'Возникла ошибка {exc}', {i})\n","    print(\"sumOfPairs = \", sumOfPairs)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"geX0-Zq6tgxU"},"source":["### Задание 1.\n","\n","Есть файл с протоколом регистраций пользователей на сайте (registrations.txt).<br>\n","Каждая строка содержит информацию о имени, электронной почте и возрасте человека. <br><br>\n","\n","\n","Надо проверить данные из файла, для каждой строки:\n"," - присутсвуют все три поля\n"," - поле имени содержит только буквы\n"," - поле email содержит @ и .\n"," - поле возраст является числом от 10 до 99<br>\n","\n","В результате проверки нужно сформировать два файла\n"," - registrations_good.log для правильных данных, записывать строки как есть\n","- registrations_bad.log для ошибочных, записывать строку и вид ошибки.<br>\n","\n","Для валидации строки данных написать метод, который может выкидывать исключения:\n"," - НЕ присутсвуют все три поля: ValueError\n"," - поле имени содержит НЕ только буквы: NotNameError (кастомное исключение)\n"," - поле email НЕ содержит @ и .(точку): NotEmailError (кастомное исключение)\n"," - поле возраст НЕ является числом от 10 до 99: ValueError\n","Вызов метода обернуть в try-except."]},{"cell_type":"code","metadata":{"id":"Eq2Ax0LntgxW","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1611578270545,"user_tz":-180,"elapsed":646,"user":{"displayName":"Андрей Щербинин","photoUrl":"","userId":"12257454032092519749"}},"outputId":"e02e6ee1-58fc-49d3-9242-bf888a9580ab"},"source":["from google.colab import drive\n","drive.mount('/content/drive')\n"],"execution_count":null,"outputs":[{"output_type":"stream","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"Arfc1e2dv65F"},"source":["\r\n","with open('/content/drive/MyDrive/SkillBox/registrations_.txt', 'r') as ff:\r\n","    try:\r\n","        good_log_file = open('/content/drive/MyDrive/SkillBox/registrations_good.log', 'w') # create / open good log file\r\n","        bad_log_file = open('/content/drive/MyDrive/SkillBox/registrations_bad.log', 'w')# create / open bad log file\r\n","        \r\n","        class NotNameError(Exception): # create NotNameError exception\r\n","            def __str__(self):\r\n","                self.name = 'NotNameError'\r\n","                return self.name\r\n","\r\n","        class NotEmailError(Exception): # create NotEmailError exception\r\n","            def __str__(self):\r\n","                self.name = 'NotEmailError'\r\n","                return self.name\r\n","\r\n","        for line in ff: # checking lines in file\r\n","            try:\r\n","                name, email, age = line.split() # checking for count of values - ValueError\r\n","                if name.isalpha() is False: # checking name consists of letters - NotNameError\r\n","                    raise NotNameError()\r\n","                elif email.rfind('@') == -1 or email.find('.') == -1: # checking email includes @ or . - NotEmailError\r\n","                    raise NotEmailError()\r\n","                age = int(age)\r\n","                if age < 10 or age > 99: # checking 10 < age < 99 - ValueError\r\n","                    raise ValueError(\"ValueError_Age\")\r\n","            except ValueError as exc:\r\n","                if 'unpack' in exc.args[0]:\r\n","                    bad_log_file.write(f'Недостаточно значений ({exc}) в {line}')\r\n","                else:\r\n","                    bad_log_file.write(f'Поле возраст НЕ является числом от 10 до 99 ({exc}) в {line}')\r\n","            except NotNameError as exc:\r\n","                bad_log_file.write(f'Поле имени содержит НЕ только буквы ({exc}) в {line}')\r\n","            except NotEmailError as exc:\r\n","                bad_log_file.write(f'Поле email НЕ содержит @ или .(точку) ({exc}) в {line}')\r\n","            else:\r\n","                good_log_file.write(line) # writing  right data to good log\r\n","    except Exception as exc:\r\n","        print(\"Неожиданная ошибка:\", exc)\r\n","    finally:\r\n","        good_log_file.close() # close good log file\r\n","        bad_log_file.close()# close bad log file"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"JMCyAWRAz9yG"},"source":[""],"execution_count":null,"outputs":[]}]}